docs/2019-12-16-Question-Mark-Variables-With-Ensures.md

```diff
@@ -0,0 +1,313 @@
+//  \forall Is:List{Int} . fresh(Is) -> * \exists I:Int . \neg( I \in Is) /\ I
+```
+
+### Frontend Syntax
+
+We use the following K syntax to define `fresh(Is)`
+
+```
+syntax Exp ::= fresh (List{Int})
+rule fresh(Is:List{Int}) => ?I:Int
+  ensures notBool (?I inList{Int} Is)
+```
+
+A variant of this would be a `choiceInList(Is)` language construct which would
+choose some number from a list:
+
+```
+syntax Exp ::= choiceInList (List{Int})
+rule choiceInList(Is:List{Int}) => ?I:Int
+  ensures ?I inList{Int} Is
+```
+
+
+## Example C: Arbitrary Number (Unspecific Function) `arb()`
+
+The function `arb()` is not a PL construct, but a mathematical function.
+Therefore, its axioms do not use rewrites but equalities.
+
+### Axioms
+
+The intended semantics of `arb()` is that it is an unspecified nullary function.
+The exact return value of `arb()` is unspecified in the semantics but up to the
+implementations.
+However, being a mathematical function, `arb()` must return the same value in
+any one implementation.
+
+The axiom of `arb()` is
+
+```
+(ARB) \exists I:Int . arb() = I
+//    this is exactly the (Function) axiom
+//    this is not equivalent to
+//      arb() = (\exists I:Int . I)
+```
+
+There are many variants of `arb()`. For example, `arbInList(Is)` is
+an unspecified function whose return value must be an element from `Is`.
+Here is its axiom
+
+```
+(ARB) \forall Is:List{Is} . \exists I:Int . I \in Is /\ arbNotInList(Is) = I
+```
+
+Note that `arbInList(Is)` is different from `choiceInList(Is)`, because
+`choiceInList(Is)` _transitions_ to an integer in `Is` (could be a different one
+each time it is used), while `arbInList(Is)` _is equal to_ a (fixed)
+integer not in `Is`.
+
+### Frontend Syntax
+
+We do not need special frontend syntax to define `arb()`.
+We only need to define it in the usual way as a function
+(instead of a language construct), and provide no axioms for it
+(the `(Function)` axiom is generated automatically from the `function`
+attribute annotation).
+
+W.r.t. the `arb` variants, we can use `?` variables and the `function`
+annotation to signal that the defining rules should be Skolemized when
+kompiled to KORE ([see below](#skolemization)) for the defining rules.
+For example, `arbInList` could be defined in K as
+
+```
+syntax Int ::= arbInList(List{Int}) [function]
+rule arbInList(Is:List{Int}) => ?I:Int
+  ensures ?I inList{Int} Is
+```
+
+### Skolemization
+
+One possible approach to implementing the above would be through
+[Skolemization](https://en.wikipedia.org/wiki/Skolem_normal_form),
+i.e., by replacing the existential variables with fresh uninterpreted functions
+applied to the universally quantified variables.
+
+For example, the `(ARB)` rule above could be Skolemized to
+
+```
+(ARB') \forall Is:List{Is} . arbI(Is) \in Is /\ arbInList(Is) = arbI(Is)
+```
+where
+```
+arbI : List{Int} -> Int
+```
+
+is a symbol satisfying the `(Function)` axiom, without further constraints.
```

Comment:
Why is it necessary to introduce an auxiliary symbol instead of using an existential quantifier as for `arb`? For example, why is this insufficient?

```
\forall Is:List{Int} . \exists I:Int . notEmpty(Is) -> arbInList(Is) = I ∧ inList(I, Is)
```

Reply:
Logically, it is sufficient, but operationally I think we are better off if Skolemizing, because it's not clear how we could instantiate this axiom if we don't.

Reply:
It's not clear to me how we could instantiate `arbI(Is)` using the `(Function)` axiom, either. We have to introduce the same variable for each occurrence, but also distinguish between different arguments.